### DISCUSSION TITLE
通过query()获取向量耗时高

### DISCUSSION THREAD

**Discussion by WLCOOLONGS** [milvus-io/milvus#21013] [Q&A and General discussion]
通过sdk query查询向量耗时高，2千万数据约500ms，同样的数据量faiss查询几ms，是配置有问题吗

**Comment by yhmo** [comment]
比faiss慢是一定的，faiss相当于数据在内存里直接计算，并且是一整块数据建的索引。
milvus有组件间以及客户端/服务端之间的传输延迟，有内部调度的消耗，可能会有两三毫秒的系统开销。milvus为了支持海量数据，是把数据分块(segment)管理的，每个segment建立独立的索引，然后分别搜索合并结果集，这里会有一些合并的消耗，并且分块建索引并没有一整块建索引搜索起来那么高效。
但是，2千万数据，索引都建好并且比较合理的情况下，耗时是可以压在10ms以下的。500ms我猜测可能是某些segment还没建好索引，采用了暴搜的方式。暴搜比索引搜索要慢至少1个量级。

**Reply by WLCOOLONGS** [reply]
获取相似向量确实可能在10ms以下，但是通过sdk中的query方法获取单个id对应的向量耗时比较高

**Reply by yhmo** [reply]
query()是通过id找到某个segment的存储位置，然后读出其中一小段(可能是几兆大小)来获取向量的数据的，所以耗时主要是在IO上。这块还没怎么优化过，目前确实性能差。
很多时候，向量的原始数据并不在内存里，因为有些索引是没有向量原始数据的

**Reply by xiaofan-luan** [reply]
我们准备在优化这一块，查询原始向量目前比较慢
